### PR TITLE
Draft rewrite of py3clean to use vanilla Python 3

### PR DESCRIPTION
**Pull request by HoECoder**
Hi,

Here is a draft of a Python 3 version using only what is in the standard Python 3 library.

I so far tested across Windows 10 and Raspbian Buster.

This is just a first draft. I know there are some things it does not yet do compared to the original. I wanted to get it up there for review before going to far.

### CODE REVIEW

**Comment by HoECoder** [comment]
Fixes #10

**Comment by bittner** [comment]
Let us rename `pyclean` module to `py2clean` and write a new implementation into a new module `pyclean`. What do you think? -- If you want I can do the renaming, and ensure the whole thing keeps working, in a separate PR.

And, most importantly, let's write tests for all new code!

**Comment by bittner** [comment]
I have added a new `pyclean.modern` including a test module for it on `master`:

- [pyclean/modern.py](https://github.com/bittner/pyclean/blob/master/pyclean/modern.py)
- [tests/test_modern.py](https://github.com/bittner/pyclean/blob/master/tests/test_modern.py)

You can start from there. Simply run `tox` or `tox -e py37` to run the tests.

**Comment by HoECoder** [comment]
Alright, I will take a look later in the week when I have some more time between work and family.

**Comment by bittner** [comment]
That seems like an easy exercise, now that you already have the implementation in your PR.

Just a few hints to keep this contribution painless:

* Keep the docstrings small and readable (remove the noisy parameter descriptions that PyCharm generates) :snake: 
* Don't use f-strings for now (we still need to support Python 3.4 and 3.5, which don't support f-strings yet) :old_key: 
* Instead of comments try to make your code self-explanatory (and sweet, and beautiful) :lollipop: 

I'm really looking forward to get this working on Windows with your help!

**Comment by bittner** [comment]
When you continue and want some support, let me know! :computer:

**Comment by bittner** [comment]
:bell: ***ping***

File: pyclean/py3clean.py [hunk 1]
```diff
@@ -2,175 +2,100 @@
...
-        if args.package:
-            files = files & pfiles
-    else:
-        files = pfiles
-
-    for filename in files:
-        d.send(filename)
+    dirs_to_check, files_to_remove, dirs_to_remove = find_files(args.root_dir)
+    print(f"Found {len(files_to_remove)} files to remove in {len(dirs_to_check)} directories. Can remove {len(dirs_to_remove)} directories.")
+    remove_files_and_dirs(files_to_remove, dirs_to_remove)
```
Comment (bittner) [inline pyclean/py3clean.py:96]: It seems like adding these three lines (plus the two methods therein, obviously) to the [modern](https://github.com/bittner/pyclean/blob/master/pyclean/modern.py) module would be sufficient.

Do you think you have time to do this in the next couple of days?

**Comment by HoECoder** [comment]
I've not forgotten about this. I have been waylaid by other items.

**Comment by HoECoder** [comment]
I put it under pyclean/modern.py.

Removed the f-strings, cleaned up some doc-strings.

I'm of the opinion that a long doc-string isn't bad. I was actually inserting the the argument stuff from Pycharm; I was going it by hand in good-ol' vi.

I also like leaving some comments in there. It is one point I disagree with in terms of the comment ethos in python. I've maintained a large enough code base of my own that even with good algorithms and names, I still benefit from some description of _what_ I was trying to do.

I'll have to read up on pytest; I'm a heavy unittest user and have little experience with pytest.

**Comment by bittner** [comment]
> I'll have to read up on pytest; I'm a heavy unittest user and have little experience with pytest.

Pytest is basically unittest with

1. functions _instead of_ classes and methods
1. `assert` _instead of_ `assertFoo...` boilerplate methods

But pytest can also run tests written in unittest. So, if you don't want to wrap your head around Pytest simply write your tests in unittest, and I'll convert them to Pytest later. Not a big deal!  (I prefer having extra work to not having tests at all.)

**Comment by bittner** [comment]
> I also like leaving some comments in there. It is one point I disagree with in terms of the comment ethos in python. I've maintained a large enough code base of my own that even with good algorithms and names, I still benefit from some description of what I was trying to do.

Try to put the information in comments into module and function docstrings. That usually makes a lot of sense (especially for the _what_).

**Comment by bittner** [comment]
Note that the linting currently fails. Try to run `tox` locally before pushing the changes. There should be only green ticks.

**Comment by HoECoder** [comment]
The problem isn't in the new code:

- https://travis-ci.org/github/bittner/pyclean/jobs/685161459#L229
- https://travis-ci.org/github/bittner/pyclean/jobs/685161459#L235
- https://travis-ci.org/github/bittner/pyclean/jobs/685161459#L237
- https://travis-ci.org/github/bittner/pyclean/jobs/685161460#L229
- https://travis-ci.org/github/bittner/pyclean/jobs/685161460#L247
- https://travis-ci.org/github/bittner/pyclean/jobs/685161460#L270

I reverted py3clean.py to the last commit from master; I've not touched py2clean.py or pypyclean.py.

The bandit errors also appear to be in these other modules. Perhaps the sensitivity of the configuration of bandit is too high?

That said, for https://travis-ci.org/github/bittner/pyclean/jobs/685161461#L239 -> https://github.com/bittner/pyclean/blob/master/pyclean/py3clean.py#L118 you should consider:

1. Not using such a broad exception as `Exception`. As I recall, the list of errors from rmdir is short:
    - I believe it is `FileNotFound` or a child of `OSError`.
3. Either re-raising the exception or consolidating a list of errors for a final user to analyze.

Having not analyzed pypyclean.py, I can't speak to its calling of `shutil` or `subprocess` and if that is a good idea or not.

**Comment by bittner** [comment]
Don't worry about the legacy code. This is original code from Debian. The [builds all pass](https://travis-ci.org/github/bittner/pyclean/builds) on `master`, because some are allowed to fail. Don't worry about that. Ticks are all green.

You only need to worry about the code you add.

**Comment by HoECoder** [comment]
That was my point, the build is not failing the code I added. It is failing in py3clean.py and py2clean.py.

image
image

image
image

**Comment by HoECoder** [comment]
If I run flake8, pylint, or bandit against the new code there are no issues:

```
redac:~/pyclean $ source venv/bin/activate
(venv) redac:~/pyclean $ bandit pyclean/modern.py
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 3.7.3
Run started:2020-05-13 14:15:52.681552

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 67
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
        Total issues (by confidence):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
Files skipped (0):
(venv) redac:~/pyclean $ flake8 pyclean/modern.py
(venv) redac:~/pyclean $ pylint pyclean/modern.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

(venv) redac:~/pyclean $
```

**Comment by bittner** [comment]
@HoECoder This PR is still a draft. Do you plan to continue and clean up up your contribution, or do you want me to merge it as it is and go from there?

I could give you more hints on what to improve, but I'm not sure if you would want to have more input from my side. Let me know how to proceed.

**Comment by HoECoder** [comment]
@bittner Honestly, I'm a bit frustrated here.

I'm being asked to make tests and linting pass in items I wasn't coding. The linting failures documented in Travis are irrelevant to the code I was writing. I get that the baseline lints out OK, but the same linting fails on the branch. And resolution of that was not something I signed up for.

It feels like the goal posts keep getting moved.

You may take the code and do as you please; I hope it is of some use and pyclean becomes truly universal. In the interim, I strongly suggest you update the package definition in [PyPI](https://pypi.org/project/pyclean/). The package today might be pure Python but it only supports Debian. And if it is only intended to support Debian (or that is the only bandwidth you have to spare), I suggest leaving that at the top of the package description as well.

I solved the problem I had back on April 10 and personally I've moved on; I've got a very stripped down version of the code here that does precisely what I needed it to do.

I've only got so much bandwidth in my personal life I have to carefully pick and choose what I spend my time on. I was willing to try to help, but it felt as is barriers to entry were erected after I wanted to try to help. I know there might be standards, but they were not clear until after the fact. It became not fun or challenging. And if it isn't fun or challenging, I'd rather not do it in my own time.

**Comment by bittner** [comment]
I'm sorry you feel that way. I can assure you I don't erect barriers. Writing tests is just necessary, but no worries: As I proposed, you are free to say I should merge your code without tests. If CI behaves confusingly I apologize, we need to understand why. Today's world is complex.

This is all just how open source contributions work. You can benefit both ways, for your fame and for learning new stuff. Especially in the beginning the learning is hard. I'm still willing to help you along, though. If you want to try to continue feel free to reopen this PR.

Thank you sincerely for trying and bringing it that far! :medal_sports:
